File: jsengine-android/jsengine/src/main/java/com/cliqz/jsengine/AntiTracking.java
package com.cliqz.jsengine;

import android.util.Log;

import com.cliqz.jsengine.v8.api.WebRequest;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

/**
 * Created by sammacbeth on 21/10/2016.
 */

public class AntiTracking {

    private final static String TAG = AntiTracking.class.getSimpleName();

    private final static String MODULE_NAME = "antitracking";
    private final static String ENABLE_PREF = "antiTrackTest";
    private final static String QSBLOCKING_PREF = "attrackRemoveQueryStringTracking";
    private final static String BLOOM_FILTER_PREF = "attrackBloomFilter";
    private final static String FORCE_BLOCK_PREF = "attrackForceBlock";

    private final Engine engine;
    private final WebRequest webRequest;

    public AntiTracking(Engine engine) {
        this.engine = engine;
        this.webRequest = engine.webRequest;
    }

    /**
     * Get the default prefs to initialise the antitracking module in enabled state
     * @return
     */
    public static Map<String, Object> getDefaultPrefs() {
        return getDefaultPrefs(true);
    }

    /**
     * Get the default prefs to initialise the antitracking module
     * @param enabled
     * @return
     */
    public static Map<String, Object> getDefaultPrefs(final boolean enabled) {
        final Map<String, Object> prefs = new HashMap<>();
        prefs.put(BLOOM_FILTER_PREF, true);
        prefs.put(ENABLE_PREF, enabled);
        prefs.put(QSBLOCKING_PREF, true);
        prefs.put(FORCE_BLOCK_PREF, true);
        return prefs;
    }

    public void setEnabled(final boolean enabled) throws ExecutionException {
        engine.setPref(BLOOM_FILTER_PREF, true);
        engine.setPref(QSBLOCKING_PREF, true);
        engine.setPref(ENABLE_PREF, true);
        engine.setPref("modules."+ MODULE_NAME + ".enabled", enabled);
    }

    public void setForceBlockEnabled(final boolean enabled) throws ExecutionException {
        engine.setPref(FORCE_BLOCK_PREF, enabled);
    }

    public JSONObject getTabBlockingInfo(final int tabId) {
        try {
            final Object blockingInfo = engine.system.callFunctionOnModuleDefault(MODULE_NAME + "/attrack", "getTabBlockingInfo", tabId, webRequest.getUrlForTab(tabId));
            return new JSONObject(blockingInfo.toString());
        } catch (ExecutionException e) {
            Log.e(TAG, "getTabBlockingInfo", e);
        } catch (JSONException e) {
            // shouldn't happen - the parsed json comes directly from JSON.stringify in JS
            throw new RuntimeException(e);
        }
        // on failure return empty object
        return new JSONObject();
    }

    public boolean isWhitelisted(final String url) {
        try {
            final Object whitelisted = engine.system.callFunctionOnModuleDefault(50, MODULE_NAME + "/attrack", "isSourceWhitelisted", url);
            return whitelisted.equals(Boolean.TRUE);
        } catch (ExecutionException e) {
            Log.e(TAG, "isWhitelisted", e);
        } catch (TimeoutException e) {
            return true;
        }
        return false;
    }

    public void addDomainToWhitelist(final String url) {
        try {
            engine.system.callVoidFunctionOnModuleAttribute(MODULE_NAME + "/attrack", new String[]{"default"}, "addSourceDomainToWhitelist", url);
        } catch (ExecutionException e) {
            Log.e(TAG, "addDomainToWhitelist", e);
        }
    }

    public void removeDomainFromWhitelist(final String url) {
        try {
            engine.system.callVoidFunctionOnModuleAttribute(MODULE_NAME + "/attrack", new String[]{"default"}, "removeSourceDomainFromWhitelist", url);
        } catch (ExecutionException e) {
            Log.e(TAG, "removeDomainFromWhitelist", e);
        }
    }

}

File: jsengine-ios/jsengine/AntiTracking.swift
//
//  AntiTracking.swift
//  jsengine
//
//  Created by Ghadir Eraisha on 1/13/17.
//  Copyright © 2017 Cliqz GmbH. All rights reserved.
//

import Foundation

public class AntiTracking {
    
    //MARK: - Constants
    private static let moduleName = "antitracking"
    private static let enablePref = "antiTrackingTest"
    private static let qsBlockingPref = "attrackRemoveQueryStringTracking"
    private static let bloomFilterPref = "attrackBloomFilter"
    private static let forceBlockPref = "attrackForceBlock"
    
    //MARK: - Instant variables
    var engine: Engine
    var webReuest: WebRequest
    
    //MARK: - Init
    public init(engine: Engine) {
        self.engine = engine
        self.webReuest = engine.webRequest!
    }
    
    //MARK: - Public APIs
    public class func getDefaultPrefs(enabled: Bool? = true) -> [String: Bool] {
        var prefs: [String:Bool] = [String:Bool]()
        prefs[AntiTracking.enablePref] = enabled
        prefs[AntiTracking.bloomFilterPref] = true
        prefs[AntiTracking.qsBlockingPref] = true
        prefs[AntiTracking.forceBlockPref] = true
        
        return prefs
    }
    
    public func setEnabled(enabled: Bool) {
        dispatch_async(self.engine.dispatchQueue) { 
            self.engine.setPref(AntiTracking.bloomFilterPref, prefValue: true)
            self.engine.setPref(AntiTracking.qsBlockingPref, prefValue: true)
            self.engine.setPref(AntiTracking.enablePref, prefValue: true)
            self.engine.setPref("modules." + AntiTracking.moduleName + ".enabled", prefValue: enabled)
        }
    }
    
    public func setForceBlockEnabled(enabled: Bool) {
        dispatch_async(self.engine.dispatchQueue) {
            self.engine.setPref(AntiTracking.forceBlockPref, prefValue: enabled)
        }
    }
    
    public func getTabBlockingInfo(tabId: Int) -> [NSObject : AnyObject]? {
        guard self.engine.isRunning() else {
            return nil
        }
        
        do {
            var argument = [AnyObject]()
            argument.append(tabId)
            if let tabUrl = webReuest.getUrlForTab(tabId) {
                argument.append(tabUrl)
            }
            let blockingInfo = try engine.systemLoader?.callFunctionOnModuleAttribute(AntiTracking.moduleName + "/attrack", attribute: ["default"], functionName: "getTabBlockingInfo", arguments: argument)
            return blockingInfo?.toDictionary()
        } catch let error as NSError {
            DebugLogger.log("<< Error in AntiTracking.getTabBlockingInfo: \(error)")
        }
        return nil
    }
    
    public func isWhitelisted(url: String) -> Bool? {
        guard self.engine.isRunning() else {
            return nil
        }
        
        do {
            let whitelisted = try engine.systemLoader?.callFunctionOnModuleAttribute(AntiTracking.moduleName + "/attrack", attribute: ["default"], functionName: "isSourceWhitelisted", arguments: [url])
            if let whitelisted =  whitelisted {
                return whitelisted.toBool()
            }
        } catch let error as NSError {
            DebugLogger.log("<< Error in AntiTracking.isWhitelisted: \(error)")
        }
        return false
    }
    
    public func addDomainToWhitelist(url: String) {
        do {
            try self.engine.systemLoader?.callFunctionOnModuleAttribute(AntiTracking.moduleName + "/attrack", attribute: ["default"], functionName: "addSourceDomainToWhitelist", arguments: [url])
        } catch let error as NSError {
            DebugLogger.log("<< Error in AntiTracking.addDomainToWhitelist: \(error)")
        }
    }
    
    public func removeDomainFromWhitelist(url: String) {
        do {
            try self.engine.systemLoader?.callFunctionOnModuleAttribute(AntiTracking.moduleName + "/attrack", attribute: ["default"], functionName: "removeSourceDomainFromWhitelist", arguments: [url])
        } catch let error as NSError {
            DebugLogger.log("<< Error in AntiTracking.removeDomainFromWhitelist: \(error)")
        }
    }
}
